File: repowire/daemon/routes/lifecycle.py
"""Lifecycle event endpoints — provider-agnostic (tmux, containers, etc.)."""

from __future__ import annotations

from fastapi import APIRouter, Depends

from repowire.daemon.auth import require_localhost
from repowire.daemon.deps import get_lifecycle_handler
from repowire.daemon.lifecycle import (
    ClientDetachedRequest,
    PaneDiedRequest,
    SessionClosedRequest,
    SessionRenamedRequest,
    WindowRenamedRequest,
)
from repowire.daemon.routes._shared import OkResponse

router = APIRouter(tags=["lifecycle"])


@router.post("/hooks/lifecycle/pane-died")
async def hook_pane_died(
    request: PaneDiedRequest,
    _: None = Depends(require_localhost),
) -> OkResponse:
    handler = get_lifecycle_handler()
    await handler.handle_pane_died(request.pane_id)
    return OkResponse()


@router.post("/hooks/lifecycle/session-closed")
async def hook_session_closed(
    request: SessionClosedRequest,
    _: None = Depends(require_localhost),
) -> OkResponse:
    handler = get_lifecycle_handler()
    await handler.handle_session_closed(request.session_name)
    return OkResponse()


@router.post("/hooks/lifecycle/session-renamed")
async def hook_session_renamed(
    request: SessionRenamedRequest,
    _: None = Depends(require_localhost),
) -> OkResponse:
    handler = get_lifecycle_handler()
    await handler.handle_session_renamed(request.new_name, request.pane_ids)
    return OkResponse()


@router.post("/hooks/lifecycle/window-renamed")
async def hook_window_renamed(
    request: WindowRenamedRequest,
    _: None = Depends(require_localhost),
) -> OkResponse:
    handler = get_lifecycle_handler()
    await handler.handle_window_renamed(
        request.session_name, request.new_name, request.pane_ids,
    )
    return OkResponse()


@router.post("/hooks/lifecycle/client-detached")
async def hook_client_detached(
    request: ClientDetachedRequest,
    _: None = Depends(require_localhost),
) -> OkResponse:
    handler = get_lifecycle_handler()
    await handler.handle_client_detached(request.session_name)
    return OkResponse()

File: daemon-go/hooks/lifecycle.go
package hooks

import (
	"os/exec"
	"strings"
)

func RunLifecycle(args []string) int {
	if len(args) == 0 {
		return 2
	}
	path := "/hooks/lifecycle/" + args[0]
	body := map[string]any{}
	switch args[0] {
	case "pane-died":
		if len(args) < 2 {
			return 2
		}
		body["pane_id"] = args[1]
	case "session-closed", "client-detached":
		if len(args) < 2 {
			return 2
		}
		body["session_name"] = args[1]
	case "session-renamed":
		if len(args) < 2 {
			return 2
		}
		body["new_name"], body["pane_ids"] = args[1], listHookPanes(true)
	case "window-renamed":
		if len(args) < 3 {
			return 2
		}
		body["new_name"], body["session_name"], body["pane_ids"] = args[1], args[2], listHookPanes(false)
	default:
		return 2
	}
	if daemonPost(path, body) == nil {
		return 1
	}
	return 0
}

func listHookPanes(session bool) []string {
	args := []string{"list-panes"}
	if session {
		args = append(args, "-s")
	}
	args = append(args, "-F", "#{pane_id}")
	out, err := exec.Command("tmux", args...).Output()
	if err != nil {
		return nil
	}
	var panes []string
	for _, pane := range strings.Fields(string(out)) {
		panes = append(panes, pane)
	}
	return panes
}
